### PR TITLE
Add new info.Info interface and add minimal package yaml parser

### PR DESCRIPTION
**Pull request by mvo5**
This branch is motivated by the need to have the "type" of a snap available in ubuntu-device-flash so that I can infer if its a kernel or an OS snap. When those are detected u-d-f needs to setup the boot variables accordingly.

It also moves pkg.File:Open into its own package to avoid circular dependencies. If you prefer that in a separate branch I can do that.

I think this change has some potential and we could use it as a start to get packageYaml out of the main "snappy" package.

### CODE REVIEW

**Comment by mvo5** [comment]
retest this please

**Comment by chipaca** [comment]
I like what you're doing here. I don't like the name `opener`, because once you've imported it's not clear that all it opens are packages :-) but I don't have a better suggestion.

**Comment by mvo5** [comment]
Thanks @chipaca ! I renamed it to snapfile.Open now and moved the pkg.File in there too (but I can easily revert that last commit if you prefer it as pkg.File).

**Comment by mvo5** [comment]
Fwiw, the integration tests shold be ok again once bugfix/oem-compat is merged into master.

**Comment by mvo5** [comment]
Once this has landed we can land a branch for ubuntu-device-flash in 16.04 that can create all-snap images (and only all-snap images).

**Comment by mvo5** [comment]
And once we have u-d-f in 16.04 with all-snap support the integration-tests can be updated to all-snap images too.

File: pkg/info/info_package_yaml.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/ubuntu-core/snappy/pkg"
+)
+
+type packageYaml struct {
+	Name    string
+	Version string
+	Type    pkg.Type
+}
+
+// SnapInfoPackageYaml implements the meta/snap.yaml data
```
Comment (chipaca) [inline pkg/info/info_package_yaml.go]: `snap.yaml` already?

**Comment by chipaca** [comment]
:+1:

File: pkg/info/info_package_yaml.go (hunk 1 above)
Comment (mvo5) [inline pkg/info/info_package_yaml.go]: Not yet, but soon :) the whole layout is anticipating snap.yaml

**Comment by niemeyer** [comment]
For the record, we discussed this online and @mvo5 is working on some changes right now.

**Comment by mvo5** [comment]
The branch is updated now based on the feedback from @niemeyer (thanks!). Its a bit long unfortunately mostly due to the renames. Feedback (very) welcome.

File: snap/file.go [hunk 2]
```diff
@@ -43,6 +38,22 @@ type File interface {
 	// NeedsMountUnit determines whether it's required to setup
 	// a mount unit for the snap when the snap is installed
 	NeedsMountUnit() bool
+
+	// Info returns information about the given snap file
+	Info() (*Info, error)
+}
+
+// Backend implements a specific snap format
+type Backend struct {
```
Comment (niemeyer) [inline snap/file.go]: Looks like this should be private for now. I'd also take the chance we have a good term for it after "file format" (e.g. "snapFormat" and "RegisterFormat", or something similar of your liking) and reserve the generic "backend" wildcard term for something else.

File: snap/info_package_yaml.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+package snap
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+)
+
+// NewFromPackageYaml creates a new info based on the given packageYaml
+func NewFromPackageYaml(yamlData []byte) (*Info, error) {
```
Comment (niemeyer) [inline snap/info_package_yaml.go]: This should probably be `snap.InfoFromPackageYaml` now that it sits under the `snap` package (it's not a "new snap").

**Comment by niemeyer** [comment]
Just a couple of trivials now.. really like the outcome, though! LGTM

**Comment by niemeyer** [comment]
This changed enough that would be good for @chipaca to have another look.

File: snap/clickdeb/deb.go [hunk 4]
```diff
@@ -48,6 +49,12 @@ var (
 	ErrMemberNotFound = errors.New("member not found")
 )
 
+func init() {
+	snap.RegisterBackend([]byte("!<arch>\ndebian"), func(path string) (snap.File, error) {
+		return Open(path)
```
Comment (chipaca) [inline snap/clickdeb/deb.go]: why isn't this
```go
	snap.RegisterBackend([]byte("!<arch>\ndebian"), Open)
```
?

File: snappy/click.go [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // -*- Mode: Go; indent-tabs-mode: t -*-
 
 /*
- * Copyright (C) 2014-2015 Canonical Ltd
+ * Copyyright (C) 2014-2015 Canonical Ltd
```
Comment (chipaca) [inline snappy/click.go]: ccooppyyrriigghhtt
also maybe +2016.

**Comment by chipaca** [comment]
:+1:

**Pull request merged**